open-models/serving/cloud_run_ollama_gemma2_rag_qa.ipynb

```diff
@@ -0,0 +1,1144 @@
+        "We can notice how RAG chain provides a more accurate and comprehensive answer than the LLM by leveraging the [source documentation](https://cloud.google.com/run/docs/overview/what-is-cloud-run). \n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "88c5f72f1981"
+      },
+      "source": [
+        "## Conclusion\n",
+        "Congratulations. Now you know how to deploy an open model to Cloud Run powered by a GPU! Specifically, you deployed a Gemma 2 model to Cloud Run with a GPU, as part of a RAG application powered by LangChain. You were able to ask answers from Gemma 2 about a documentation page.\n",
+        "\n",
+        "For more information about your identity tokens expiring and how to refresh your tokens, see the next section below \"Appendix: Handling Identity Token Expiration\".\n",
+        "\n",
+        "To clean up the resources you created in this section, see the section at the bottom \"Cleaning up\"."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "A5VPH4gP3igf"
+      },
+      "source": [
+        "## Appendix: Handling Identity Token Expiration\n",
+        "\n",
+        "When deploying a Generative AI application Google Cloud Run, you'll often need to authenticate your requests using Identity Tokens.\n",
+        "\n",
+        "These tokens will expire hourly, requiring a mechanism for automatic refresh to ensure uninterrupted operation.\n",
+        "\n",
+        "The following helper classes provide an example of how to deal with  token refresh. It leverages the `google.auth` library to handle the authentication process and automatically refresh the token when necessary.\n",
+        "\n",
+        "\n",
+        "See the following resources for more information on authentication:\n",
+        "* [Identity Token Overview](https://cloud.google.com/docs/authentication/token-types#id)\n",
+        "* [Google Cloud Run Authentication Documentation](https://cloud.google.com/run/docs/authenticating/overview)\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "TJlxAZKz-tjY"
+      },
+      "outputs": [],
+      "source": [
+        "import time\n",
+        "\n",
+        "import google.auth\n",
+        "from google.auth.exceptions import DefaultCredentialsError\n",
+        "import google.auth.transport.requests\n",
+        "import google.oauth2.id_token\n",
+        "from pydantic.v1 import Extra\n",
+        "\n",
+        "\n",
+        "class TokenManager:\n",
+        "    def __init__(\n",
+        "        self, url=None, token_lifetime=3600\n",
+        "    ):  # Default token lifetime of 1 hour\n",
+        "        self.token = None\n",
+        "        self.expiry_time = 0\n",
+        "        self.token_lifetime = token_lifetime\n",
+        "        self.url = url\n",
+        "        self.creds, _ = google.auth.default()\n",
```

Comment:
Is there a reason to create a custom class for this? Is there a class/protocol to inherit from in the Ollama library that we can use?

Reply:
there was, as ChatOllama only supported static headers. I just had a look now and we can now pass `requests` [auth callables](https://api.python.langchain.com/en/latest/chat_models/langchain_community.chat_models.ollama.ChatOllama.html#langchain_community.chat_models.ollama.ChatOllama.auth). Thanks for the suggestion, I updated that section!

https://requests.readthedocs.io/en/latest/user/authentication/